pkg/connection/store/kubernetes/store.go

```diff
@@ -0,0 +1,168 @@
+		// No KubernetesSecretStoreConfig provided, local API Server will be
+		// used as Secret Store.
+		return &SecretStore{
+			client: local,
+			applicator: resource.NewApplicatorWithRetry(resource.NewAPIPatchingApplicator(local),
+				resource.IsAPIErrorWrapped, nil),
+			defaultNamespace: cfg.DefaultScope,
+		}, nil
+	}
+
+	// Configure client for an external API server with a given Kubeconfig.
+	kfg, err := resource.CommonCredentialExtractor(ctx, cfg.Kubernetes.Auth.Source, local, cfg.Kubernetes.Auth.CommonCredentialSelectors)
+	if err != nil {
+		return nil, errors.Wrap(err, errExtractKubernetesAuthCreds)
+	}
+	remote, err := clientForKubeconfig(kfg)
+	if err != nil {
+		return nil, errors.Wrap(err, errExtractKubernetesAuthCreds)
+	}
+
+	return &SecretStore{
+		client: remote,
+		applicator: resource.NewApplicatorWithRetry(resource.NewAPIPatchingApplicator(remote),
+			resource.IsAPIErrorWrapped, nil),
+		defaultNamespace: cfg.DefaultScope,
+	}, nil
+}
+
+// ReadKeyValues reads and returns key value pairs for a given Kubernetes Secret.
+func (ss *SecretStore) ReadKeyValues(ctx context.Context, i store.Secret) (store.KeyValues, error) {
+	s := &corev1.Secret{}
+	return s.Data, errors.Wrapf(ss.client.Get(ctx, types.NamespacedName{Name: i.Name, Namespace: ss.namespaceForSecret(i)}, s), errGetSecret)
+}
+
+// WriteKeyValues writes key value pairs to a given Kubernetes Secret.
+func (ss *SecretStore) WriteKeyValues(ctx context.Context, i store.Secret, kv store.KeyValues) error {
+	meta := secretMetadata{}
+	if len(i.Metadata) > 0 {
+		if err := json.Unmarshal(i.Metadata, &meta); err != nil {
+			return errors.Wrap(err, errParseMetadata)
+		}
+	}
+
+	t := resource.SecretTypeConnection
+	if meta.Type != "" {
+		t = meta.Type
+	}
+
+	s := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        i.Name,
+			Namespace:   ss.namespaceForSecret(i),
+			Labels:      meta.Labels,
+			Annotations: meta.Annotations,
+		},
+		Type: t,
+		Data: kv,
+	}
+
+	return errors.Wrap(ss.applicator.Apply(ctx, s), errCreateOrUpdateSecret)
+}
+
+// DeleteKeyValues delete key value pairs from a given Kubernetes Secret.
+// If no kv specified, the whole secret instance is deleted.
+// If kv specified, those would be deleted and secret instance will be deleted
+// only if there is no data left.
+func (ss *SecretStore) DeleteKeyValues(ctx context.Context, i store.Secret, kv store.KeyValues) error {
+	// NOTE(turkenh): DeleteKeyValues method wouldn't need to do anything if we
+	// have used owner references similar to existing implementation. However,
+	// this wouldn't work if the K8s API is not the same as where SecretOwner
+	// object lives, i.e. a remote cluster.
+	// Considering there is not much additional value with deletion via garbage
+	// collection in this specific case other than one less API call during
+	// deletion, I opted for unifying both instead of adding conditional logic
+	// like add owner references if not remote and not call delete etc.
```

Comment:
I agree with your comment on simplification, but it does seem like having an owner reference would be handy for relationship tracking and accounting. One option would be to have an owner ref that isn't a controller (which iirc won't trigger garbage collection). We could also just add a controller ref, and let our delete call race the garbage collector. 🤷‍♂️

Reply:
AFAIK, owner references would cause garbage collection no matter it is a controller or not. 
I am wondering if we could use some labeling here instead? Which would also be more convenient than owner references for querying/listing.

Reply:
> AFAIK, owner references would cause garbage collection no matter it is a controller or not.

Ah yes, you're right. I was misremembering how this works. I personally tend to prefer owner refs over labels for relationships where possible. Another option would be to add the owner ref, but delete it when the parent was deleted. That might be getting into diminishing returns though.

Reply:
Owner references would only work for local Kubernetes but given it is the most common case, we can still make an exception there if you prefer. This would also cover edge cases like you described [here](https://github.com/crossplane/crossplane-runtime/pull/321#discussion_r814361342). 

For other cases like remote Kubernetes, Vault, AWS Secret Manager, and so on, we may still want to have a way to keep track of which resource own which connection secret and only options seem to leverage labels/tags, etc.

Reply:
> Owner references would only work for local Kubernetes but given it is the most common case, we can still make an exception there if you prefer.

I think I do, but it doesn't need to block this PR.